python/ray/tune/impl/tuner_internal.py

```diff
@@ -191,8 +198,77 @@ def _maybe_warn_resource_contention(self):
                 stacklevel=4,
             )
 
+    def _validate_overwrite_trainable(self, original_trainable, overwrite_trainable):
```

Comment:
New errors that will be raised in certain situations relating to `overwrite_trainable`:

1. General sanity checks to guard against overwriting with an arbitrary Trainable
```python
def train_func(config):
	pass

# Initial training run
tuner = Tuner(train_func)
tuner.fit

tuner = Tuner.restore(path, overwrite_trainable=train_func)
# Log a warning (overwriting the trainable has no guarantees of restore succeeding)

def train_func_2(config):
	pass

tuner = Tuner.restore(path, overwrite_trainable=train_func_2)
# Throw an error -> must pass in the same trainable

tuner = Tuner.restore(path, overwrite_trainable=MyTrainable)
# Throw an error -> must pass in the same trainable
```

2. Making sure that a new trainable is specified on restore when using `with_parameters`

```python
# Initial training run
params = ...
tuner = Tuner(tune.with_parameters(train_func, **params))
tuner.fit()

tuner = Tuner.restore(path)
tuner.fit()
# Throw an error, tell user that they need to re-specify trainable with params to continue training

tuner = Tuner.restore(path, overwrite_trainable=tune.with_parameters(train_func, **params))
# Log a warning (overwriting the trainable has no guarantees of restore succeeding)
```